src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+package io.strimzi.kafka.bridge.tracing;
+
+import io.jaegertracing.Configuration;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import io.opentelemetry.instrumentation.api.instrumenter.messaging.MessageOperation;
+import io.opentelemetry.instrumentation.kafkaclients.TracingProducerInterceptor;
+import io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdk;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import io.strimzi.kafka.bridge.config.BridgeConfig;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.kafka.client.consumer.KafkaConsumerRecord;
+import io.vertx.kafka.client.producer.KafkaProducerRecord;
+import org.apache.kafka.clients.producer.ProducerConfig;
+
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ExecutorService;
+
+import static io.strimzi.kafka.bridge.tracing.TracingConstants.COMPONENT;
+import static io.strimzi.kafka.bridge.tracing.TracingConstants.JAEGER;
+import static io.strimzi.kafka.bridge.tracing.TracingConstants.KAFKA_SERVICE;
+import static io.strimzi.kafka.bridge.tracing.TracingConstants.OPENTELEMETRY_SERVICE_NAME_ENV_KEY;
+import static io.strimzi.kafka.bridge.tracing.TracingConstants.OPENTELEMETRY_SERVICE_NAME_PROPERTY_KEY;
+import static io.strimzi.kafka.bridge.tracing.TracingConstants.OPENTELEMETRY_TRACES_EXPORTER_ENV_KEY;
+import static io.strimzi.kafka.bridge.tracing.TracingConstants.OPENTELEMETRY_TRACES_EXPORTER_PROPERTY_KEY;
+
+/**
+ * OpenTelemetry implementation of Tracing.
+ */
+class OpenTelemetryHandle implements TracingHandle {
+
+    private Tracer tracer;
+    private ExecutorService service;
```

Comment:
what's the reason for storing the `ExecutorService` instance here? I see it's never being used but just set.

Reply:
It's been set, and used.

Reply:
It's returned on every `BridgeExecutorServiceFactory::delegate()` call ...

Reply:
I see now but ... I don't think that the `ExecutorService` instance belongs to a telemetry handle.
It should be created one time in the application (based on tracing enabled or not and which one) and provided to Vert.x.
It would avoid to have this `delegate()` stuff called every time.
The main problem is that we know about if tracing is enabled through the `ConfigRetriever` ...

```java
ConfigRetriever retriever = ConfigRetriever.create(vertx, options);
```

where we already need the Vert.x instance.
We should know about telemetry setting before that and getting the `ExecutorService` accordingly. 
I am thinking how to get a good solution for that ... unless you already have one :-)

Reply:
Yeah, that's the main problem and reason for that `delegate()`.
Otoh, those calls are not costly, they simply return existing instances, no additional computation.
No new idea for a better solution atm unfortunately. :-)

Reply:
So ... this kind of solution to instrument Kafka could be fine with Vert.x but when using Kafka Java native client you don't have access to the executor, executor service or thread where the interceptor onSend is executed.
Thinking in this way, I had an idea about avoiding the executor service wrapping but doing a step back with a custom interceptor (which is anyway the only way you would have with Kafka Java native).
The implementation would be this one ...

```java
public static class ContextAwareTracingProducerInterceptor<K, V> extends TracingProducerInterceptor<K, V> {
        @Override
        public ProducerRecord<K, V> onSend(ProducerRecord<K, V> record) {
            return Context.current().wrapFunction(super::onSend).apply(record);
        }
    }
}
```

Just really simple. It wraps the `onSend` of the default producer interceptor and what the `wrapFunction` does is simple as well:

```java
default <T, U> Function<T, U> wrapFunction(Function<T, U> function) {
    return t -> {
      try (Scope ignored = makeCurrent()) {
        return function.apply(t);
      }
    };
  }
```

So it's kind of something you did in the previous implementation.
The only thing is that I have to debug more because it's not working in the sense that the HTTP span and Kafka span are not related together when sending :-(
Can you give it a try as well? Maybe there is something I am doing wrong and I am blind to see it?
This solution would avoid to inject an executor service and would be a kind of pattern for Kafka producer instrumentation (when the native Java client is used)

Reply:
I guess it's not worked AS EXPECTED :-) ... because in the custom interceptor we are already in the other thread in the Kafka client so the context is different and empty ... it's too late :-(

Reply:
Vert.x seems to have a TP support for tracing here https://github.com/eclipse-vertx/vertx-tracing
Of course refactoring everything is not feasible and the feature is TP but what we should try is using the `VertxContextStorageProvider` instead of the default one which is based on thread local.

https://github.com/eclipse-vertx/vertx-tracing/blob/master/vertx-opentelemetry/src/main/java/io/vertx/tracing/opentelemetry/VertxContextStorageProvider.java

This Vert.x storage seems to be able to correlate Vert.x context with OpenTelemetry context and that could help.
Right now I am not able to set this storage yet :-(

Reply:
> Right now I am not able to set this storage yet :-(

What's the problem?
Or should I have a look at it?

Reply:
VCSP is automatically found -- via META-INF/services.
But ...

```
Exception in thread "vert.x-internal-blocking-2" java.lang.NullPointerException: Cannot invoke "io.vertx.core.Context.getLocal(Object)" because "vertxCtx" is null
	at io.vertx.tracing.opentelemetry.VertxContextStorageProvider$VertxContextStorage.attach(VertxContextStorageProvider.java:27)
	at io.vertx.tracing.opentelemetry.VertxContextStorageProvider$VertxContextStorage.attach(VertxContextStorageProvider.java:23)
	at io.opentelemetry.context.Context.makeCurrent(Context.java:202)
	at io.opentelemetry.context.Context.lambda$wrap$1(Context.java:211)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:831)
```

Missing something more ... ?

Reply:
My problem was adding the `vertx-tracing` dep but not noticing that it was there already as `test` scope which was overriding mine. Now it seems to work :-)